File: problems/leetcode/0500/525.cpp
// 525. Contiguous Array
// https://leetcode.com/problems/contiguous-array/

using namespace std;

#include <unordered_map>
#include <vector>

class Solution {
public:
  int findMaxLength(vector<int> &nums) {
    unordered_map<int, int> recentIndex;
    recentIndex[0] = -1;

    int maxLength = 0, numberOfMoreOne = 0;
    for (int i = 0; i < nums.size(); i++) {
      numberOfMoreOne += (nums[i] == 1 ? 1 : -1);
      if (recentIndex.find(numberOfMoreOne) != recentIndex.end()) {
        maxLength = max(maxLength, i - recentIndex[numberOfMoreOne]);
      } else {
        recentIndex[numberOfMoreOne] = i;
      }
    }

    return maxLength;
  }
};

File: problems/leetcode/0500/525/525.py
# 525. Contiguous Array
# https://leetcode.com/problems/contiguous-array/description/

from typing import *

class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        contiguousLength = float("-inf")

        differences = dict()
        differences[0] = -1

        counts = [0, 0]
        for i, num in enumerate(nums):
            counts[num] += 1
            diff = counts[1] - counts[0]

            if diff in differences:
                contiguousLength = max(contiguousLength, i - differences[diff])
            else:
                differences[diff] = i

        return 0 if contiguousLength == float("-inf") else contiguousLength
